hanish/bot.py

```diff
@@ -0,0 +1,409 @@
+from .utils import memoized
+from .darksky import DarkSky
+from .zipcode import ZipCodeDB
+from slackclient import SlackClient
+
+# time to wait between websocket reads
+POLLING_INTERVAL = 1
+
+
+##########################################################################
+## Helper Functions
+##########################################################################
+
+def environ_default(kwargs, name, envvar, default=None, required=False):
+    """
+    This helper function fetches a default value from the specified envvar if
+    the name doesn't exist in the kwargs dictionary. If the name or envvar are
+    not defined, then this function returns the default value unles required
+    is True, at which point an ImproperlyConfigured exception is raised.
+    """
+
+    # If the name is in the keyword arguments, return its value
+    if name in kwargs: return kwargs[name]
+
+    # If the envvar is in the environment, return its value.
+    if envvar in os.environ: return os.environ[envvar]
+
+    # If required is True and we're here, raise an exception.
+    if required:
+        raise ImproperlyConfigured(
+            "Required config {} not found in ${}".format(name, envvar)
+        )
+
+    # Otherwise return the default
+    return default
+
+
+##########################################################################
+## Hanish Bot
+##########################################################################
+
+class Bot(object):
+    """
+    A Bot encapsulates the behavior and functionality of Hanish slackbot,
+    allowing it to query the Dark Sky weather service, lookup geo-coordinates
+    by zip code and to listen for and respond to messages from users on Slack.
+
+    The Bot is configured primarily from environment variables, though every
+    configuration parameter can be overriden in the constructor. However, for
+    normal usage, it is preferred that configuration is stored in the env.
+
+    Parameters
+    ----------
+    name: string, environment: $SLACK_BOT_NAME
+        The name of the slackbot identified in Slack chat.
+
+    slack_api_key: string, environment: $SLACK_ACCESS_TOKEN
+        The api token to access the Slack API, starts with "xoxb-"
+
+    darksky_api_key: string, environment: $DARKSKY_ACCESS_TOKEN
+        The api token to access the Dark Sky API
+
+    default_zip: string, environment: $DEFAULT_ZIP_CODE
+        The default zip code to use when making lookups
+
+    zipcodes: string, environment: $ZIPCODE_DATABASE
+        Path to the zip codes database CSV file.
+    """
+
+    def __init__(self, **config):
+        # Holds the weather for yesterday to check if things have changed.
+        self.yesterday = None
+
+        # NOTE: Normally PEP8 requires breaks at 78 chars, but longer here so
+        # that I can more quickly see the configurations and variable names.
+        # Get app properties from the configuration.
+        self.name = environ_default(config, "name", "SLACK_BOT_NAME", "hanish")
```

Comment:
I completely understand your comments @rbalsano-ci  -- and actually my preference is similar, but to use a mixture of YAML configs for sane defaults at the system, user, and local level, as well as the environment for secure variables, then allow manipulations at instantiation for testing. In fact, I wrote a library, [confire](https://github.com/bbengfort/confire) to handle such configuration situations exactly, but I didn't want to inject that level of complexity in the prototype.